--- a/ftp_list.php	2019-06-20 06:54:05.000000000 +0700
+++ b/ftp_list.php	2019-07-03 10:32:42.193161086 +0700
@@ -1,4 +1,13 @@
 #!/bin/sh
+# ===========================================================
+# Original script was written by Directadmin
+# ===========================================================
+# Patched:
+#  sFTP/SSH support added
+#  By Alex Grebenschikov, Poralix, www.poralix.com
+#  Last modified: Tue Jan 30 12:43:50 +07 2018
+#  Version: 0.1.poralix $ Tue Jan 30 12:43:50 +07 2018
+# ===========================================================
 
 FTPLS=/usr/bin/ncftpls
 CURL=/usr/local/bin/curl
@@ -29,6 +38,70 @@
 touch $DUMP
 chmod 600 $DUMP
 
+# Poralix
+SSH_SFTP="/usr/bin/sftp";
+[ -x "${SSH_SFTP}" ] || SSH_SFTP="/bin/sftp";
+SSHPASS="/usr/bin/sshpass";
+[ -x "${SSHPASS}" ] || SSHPASS="/bin/sshpass";
+USE_SSH=0;
+SSH_PORTS="22 2200 22022";
+SSH_KEY="";
+for SSH_PORT in ${SSH_PORTS};
+do
+	# Working with SSH?
+	if [ -n "${PORT}" ] && [ "${PORT}" = "${SSH_PORT}" ]; then
+		USE_SSH=1;
+		break;
+	fi;
+done;
+
+list_files_ssh()
+{
+	# All the binaries installed?
+	if [ ! -x "${SSH_SFTP}" ] || [ ! -x "${SSHPASS}" ]; then
+		echo "";
+		echo "*** Backup not uploaded ***";
+		echo "Can not find SFTP and/or SSHPASS binaries";
+		echo "";
+		echo "Please install sftp and/or sshpass using OS repository";
+		echo "";
+		exit 10;
+	fi;
+
+	# Use SSH KEY?
+	if [ -n "${ftp_password}" ] && [ -f "${ftp_password}" ]; then
+		ssh-keygen -P '' -e -y -f "${ftp_password}" >/dev/null 2>&1;
+		if [ "$?" = "0" ]; then
+			SSH_KEY="${ftp_password}";
+		fi;
+	fi;
+
+	TMP_BATCH_FILE=$(mktemp);
+	#echo "progress" >> ${TMP_BATCH_FILE};
+	echo "cd ${ftp_path}" >> ${TMP_BATCH_FILE};
+	echo "ls -1" >> ${TMP_BATCH_FILE};
+	echo "quit" >> ${TMP_BATCH_FILE};
+
+	if [ -z "${SSH_KEY}" ]; then
+		echo "${ftp_password}" >> ${CFG};
+		${SSHPASS} -f${CFG} ${SSH_SFTP} -C -oBatchMode=no -b ${TMP_BATCH_FILE} -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no -oPort=${PORT} ${ftp_username}@${ftp_ip} > ${DUMP} 2>&1;
+	else
+		${SSH_SFTP} -C -oBatchMode=no -b ${TMP_BATCH_FILE} -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no -oPort=${PORT} -oIdentityFile=${SSH_KEY} ${ftp_username}@${ftp_ip} > ${DUMP} 2>&1;
+	fi;
+	RET=$?;
+
+	[ -f "${TMP_BATCH_FILE}" ] && rm -f ${TMP_BATCH_FILE};
+
+	if [ "${RET}" -ne 0 ]; then
+		echo "[list] sftp return code: $RET";
+		cat ${DUMP};
+	else
+		cat ${DUMP} | grep -v '^sftp> ' | grep -E '^(admin|reseller|user)\.(.*)\.tar(|\.gz)$';
+	fi
+}
+# Poralix
+
+
 #######################################################
 # FTP
 list_files()
@@ -122,18 +195,23 @@
 		COLS=`awk '{print NF; exit}' $DUMP`
 		cat $DUMP | grep -v -e '^d' | awk "{ print \$${COLS}; }"
 	fi
+
+
 }
 
 
 #######################################################
 # Start
 
-if [ "${FTPS}" = "1" ]; then
-	list_files_ftps
+if [ "${USE_SSH}" = "1" ]; then
+	list_files_ssh;
 else
-	list_files
-fi
-
+	if [ "${FTPS}" = "1" ]; then
+		list_files_ftps
+	else
+		list_files
+	fi
+fi;
 
 rm -f $CFG
 rm -f $DUMP
